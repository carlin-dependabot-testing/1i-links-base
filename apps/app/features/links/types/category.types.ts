export interface Category {
  name: string
  icon?: string
  description?: string
  timestamp?: string
  linkCount?: number
}
