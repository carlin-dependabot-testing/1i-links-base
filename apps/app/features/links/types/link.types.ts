import type { LinksApp } from '@/features/links/schemas/link.schema'

export type { LinksApp }

export interface ClickData {
  count: number
  url: string
  environment?: string
  lastClicked?: string
}

export interface LinkClickStore {
  clickCounts: Record<string, ClickData>
  incrementClickCount: (
    linkId: string,
    linkData: { url: string; environment?: string }
  ) => void
  getClickCount: (linkId: string) => number
  resetClickCounts: () => void
}
