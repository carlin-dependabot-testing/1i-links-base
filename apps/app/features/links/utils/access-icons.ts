import { type AccessType } from '../constants/access-types'

const ACCESS_TYPE_URLS = {
  public: 'https://www.google.com',
  okta: 'https://www.okta.com',
  ldap: 'https://www.openldap.org',
  google: 'https://accounts.google.com',
  basic: 'https://www.w3.org',
  custom: 'https://www.oauth.net'
} as const

export const getAccessTypeIconUrl = (accessType: AccessType): string => {
  const baseUrl = ACCESS_TYPE_URLS[accessType]
  return `https://www.google.com/s2/favicons?sz=32&domain_url=${encodeURIComponent(baseUrl)}`
}
