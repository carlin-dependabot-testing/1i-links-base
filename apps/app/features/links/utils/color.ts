import { formatHex, rgb } from 'culori'

import { DEFAULT_COLOR } from '@/constants'

/**
 * Extracts the dominant color from an image URL
 */
export const extractColorFromImage = async (
  imageUrl: string
): Promise<string> => {
  return new Promise(resolve => {
    const img = document.createElement('img')
    img.crossOrigin = 'anonymous'

    // Add timeout to prevent hanging
    const timeout = setTimeout(() => {
      console.error('Image loading timeout')
      resolve(DEFAULT_COLOR)
    }, 5000)

    img.onload = () => {
      clearTimeout(timeout)

      try {
        // Test if we can actually access the image data
        const canvas = document.createElement('canvas')
        const ctx = canvas.getContext('2d', { willReadFrequently: true })

        if (!ctx) {
          console.error('Could not get canvas context')
          resolve(DEFAULT_COLOR)
          return
        }

        canvas.width = img.width
        canvas.height = img.height

        // This will throw if there's a CORS issue
        ctx.drawImage(img, 0, 0)
        try {
          // This will throw if we can't access the pixel data
          ctx.getImageData(0, 0, 1, 1)
        } catch (e) {
          console.error('CORS error: Cannot access image data', e)
          resolve(DEFAULT_COLOR)
          return
        }

        // Now proceed with the actual color extraction
        const maxDimension = 100
        const scale = Math.min(
          1,
          maxDimension / Math.max(img.width, img.height)
        )

        canvas.width = img.width * scale
        canvas.height = img.height * scale

        ctx.drawImage(img, 0, 0, canvas.width, canvas.height)
        const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height)
        const data = imageData.data

        let r = 0,
          g = 0,
          b = 0
        let count = 0

        for (let i = 0; i < data.length; i += 4) {
          if (data[i + 3] < 128) continue // Skip transparent pixels
          r += data[i]
          g += data[i + 1]
          b += data[i + 2]
          count++
        }

        if (count > 0) {
          r = Math.round(r / count)
          g = Math.round(g / count)
          b = Math.round(b / count)

          const color = formatHex(
            rgb({ mode: 'rgb', r: r / 255, g: g / 255, b: b / 255 })
          )
          if (color) {
            resolve(color)
          } else {
            console.error('Could not format color')
            resolve(DEFAULT_COLOR)
          }
        } else {
          console.error('No valid pixels found')
          resolve(DEFAULT_COLOR)
        }
      } catch (error) {
        console.error('Error processing image:', error)
        resolve(DEFAULT_COLOR)
      }
    }

    img.onerror = () => {
      clearTimeout(timeout)
      // Try without crossOrigin as fallback
      const fallbackImg = new Image()
      fallbackImg.onload = () => {
        try {
          const canvas = document.createElement('canvas')
          const ctx = canvas.getContext('2d', { willReadFrequently: true })
          if (!ctx) {
            resolve(DEFAULT_COLOR)
            return
          }
          canvas.width = fallbackImg.width
          canvas.height = fallbackImg.height
          ctx.drawImage(fallbackImg, 0, 0)
          resolve(DEFAULT_COLOR) // Even if we can load it, we can't access the data
        } catch (e) {
          console.error('Fallback image processing failed:', e)
          resolve(DEFAULT_COLOR)
        }
      }
      fallbackImg.onerror = () => {
        console.error('Both CORS and non-CORS image loading failed')
        resolve(DEFAULT_COLOR)
      }
      fallbackImg.src = imageUrl
    }

    try {
      img.src = imageUrl
    } catch (error) {
      console.error('Error setting image source:', error)
      resolve(DEFAULT_COLOR)
    }
  })
}
