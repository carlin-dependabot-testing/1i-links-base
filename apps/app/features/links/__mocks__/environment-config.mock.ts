import { type EnvironmentConfig } from '@/features/links/types/environment.types'

export const mockEnvironmentConfig: EnvironmentConfig = {
  environmentColors: {
    production: '#0f0',
    staging: '#ff0',
    integration: '#00f'
  },
  environmentLabels: {
    production: 'Production',
    staging: 'Staging',
    integration: 'Integration'
  },
  enabledEnvironments: ['production', 'staging', 'integration'],
  environmentOrder: ['production', 'staging', 'integration']
}
