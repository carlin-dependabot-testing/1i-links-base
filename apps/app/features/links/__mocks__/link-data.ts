import { type LinksApp } from '@/features/links/types/link.types'

export const mockLinkData: Record<string, LinksApp> = {
  basic: {
    name: 'Test App',
    description: 'Test Description',
    category: 'Category 1',
    color: '#000000',
    environments: {
      production: 'https://test.com'
    },
    tags: ['test-tag'],
    createdAt: '2024-03-20',
    timestamp: '2024-03-20'
  },
  withMultipleEnvironments: {
    name: 'Multi Env App',
    description: 'App with multiple environments',
    category: 'Category 2',
    color: '#000000',
    environments: {
      production: 'https://prod.test.com',
      staging: 'https://staging.test.com',
      integration: 'https://dev.test.com'
    },
    tags: ['test-tag'],
    createdAt: '2024-03-20',
    timestamp: '2024-03-20'
  }
} as const
