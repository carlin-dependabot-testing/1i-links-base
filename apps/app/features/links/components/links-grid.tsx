'use client'

import { motion } from 'motion/react'

import { LINKS_GRID_CLASSES } from '@/constants'
import { type LinksApp } from '@/features/links/types/link.types'

import { LinkCard } from './link-card'

export interface LinksGridProps {
  links: LinksApp[]
  handleLinkClick: (link: LinksApp) => void
  clickedLinks: string[]
  handleAddToFavorites: (link: LinksApp) => void
  isFavourite: (name: string) => boolean
  isArchived: (name: string) => boolean
  onArchiveToggle: (name: string) => void
  withAnimation?: boolean
  showManageOptions?: boolean
}

export function LinksGrid({
  links,
  handleLinkClick,
  clickedLinks,
  handleAddToFavorites,
  isFavourite,
  isArchived,
  onArchiveToggle,
  withAnimation = true,
  showManageOptions = true
}: LinksGridProps) {
  return (
    <div className="@container">
      <ul className={LINKS_GRID_CLASSES}>
        {links.map((link, index) => {
          const LinkWrapper = withAnimation ? motion.li : 'li'
          const motionProps = withAnimation
            ? {
                initial: { opacity: 0, y: 20 },
                animate: { opacity: 1, y: 0 },
                exit: { opacity: 0, y: -20 },
                transition: { delay: index * 0.1 },
                className: 'h-full'
              }
            : { className: 'h-full' }

          return (
            <LinkWrapper key={link.name} {...motionProps}>
              <LinkCard
                className="h-full"
                app={link}
                handleLinkClick={() => handleLinkClick(link)}
                clickedLinks={clickedLinks}
                handleAddToFavorites={() => handleAddToFavorites(link)}
                isFavourite={isFavourite(link.name)}
                isArchived={isArchived(link.name)}
                onArchiveToggle={() => onArchiveToggle(link.name)}
                showManageOptions={showManageOptions}
              />
            </LinkWrapper>
          )
        })}
      </ul>
    </div>
  )
}
