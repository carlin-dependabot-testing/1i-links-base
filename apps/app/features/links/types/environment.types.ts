import { type Environment } from '@/features/environment/types/environment.types'

export interface EnvironmentConfig {
  enabledEnvironments: Environment[]
  environmentLabels: Record<Environment, string>
  environmentColors: Record<Environment, string>
  environmentOrder: Environment[]
}
