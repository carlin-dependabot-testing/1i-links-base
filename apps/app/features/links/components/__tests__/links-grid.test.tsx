import { default as userEvent } from '@testing-library/user-event'

import { render, screen } from '@/utils/test-utils'

import { mockLinkData } from '../../__mocks__/link-data'
import { type LinksApp } from '../../types/link.types'
import { LinksGrid } from '../links-grid'

// Create an array of mock links using the existing mock data
const mockLinks: LinksApp[] = [
  mockLinkData.basic,
  mockLinkData.withMultipleEnvironments
]

const renderComponent = (
  props: Partial<Parameters<typeof LinksGrid>[0]> = {}
) => {
  const defaultProps = {
    links: mockLinks,
    handleLinkClick: jest.fn(),
    clickedLinks: [],
    handleAddToFavorites: jest.fn(),
    isFavourite: jest.fn(),
    isArchived: jest.fn(),
    onArchiveToggle: jest.fn(),
    withAnimation: false
  }

  return render(<LinksGrid {...defaultProps} {...props} />)
}

describe('LinksGrid', () => {
  it.skip('should render correctly', () => {
    const { container } = renderComponent()

    expect(container).toMatchSnapshot()
  })

  it.skip('should render all links', () => {
    renderComponent()
    mockLinks.forEach(link => {
      expect(screen.getByText(link.name)).toBeInTheDocument()
    })
  })

  it.skip('should render without animation wrapper when withAnimation is false', () => {
    renderComponent({ withAnimation: false })
    const animatedElements = document.querySelectorAll('[data-motion-initial]')
    expect(animatedElements).toHaveLength(0)
  })

  it.skip('should call handleLinkClick when a link is clicked', () => {
    const handleLinkClick = jest.fn()
    renderComponent({ handleLinkClick })

    const firstLink = screen.getByText(mockLinks[0].name)
    firstLink.click()

    expect(handleLinkClick).toHaveBeenCalledWith(mockLinks[0])
  })

  it.skip('should call handleAddToFavorites when favorite button is clicked', async () => {
    const user = userEvent.setup()
    const handleAddToFavorites = jest.fn()

    renderComponent({ handleAddToFavorites })

    const moreInformationButton = screen.getAllByRole('button', {
      name: /Open (more menu|menu)/i
    })
    await user.click(moreInformationButton[0])

    const menu = screen.getByRole('menu')
    expect(menu).toBeInTheDocument()

    const favoriteButtons = screen.getAllByRole('menuitem', { name: /Pin/i })
    favoriteButtons[0].click()

    expect(handleAddToFavorites).toHaveBeenCalledWith(mockLinks[0])
  })

  it.skip('should call onArchiveToggle when archive button is clicked', async () => {
    const user = userEvent.setup()
    const onArchiveToggle = jest.fn()

    renderComponent({ onArchiveToggle })

    const moreInformationButton = screen.getAllByRole('button', {
      name: /Open (more menu|menu)/i
    })
    await user.click(moreInformationButton[0])

    const menu = screen.getByRole('menu')
    expect(menu).toBeInTheDocument()

    const archiveButtons = screen.getAllByRole('menuitem', { name: /Archive/i })

    await user.click(archiveButtons[0])

    expect(onArchiveToggle).toHaveBeenCalledWith(mockLinks[0].name)
  })
})
