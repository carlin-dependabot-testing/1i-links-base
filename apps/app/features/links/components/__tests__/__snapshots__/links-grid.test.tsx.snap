// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinksGrid should render correctly 1`] = `
<div>
  <script>
    !function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&true)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}if(e==='light'||e==='dark')d.style.colorScheme=e}catch(e){}}()
  </script>
  <div
    class="group/sidebar-wrapper flex min-h-svh w-full has-[[data-variant=inset]]:bg-sidebar"
    style="--sidebar-width: 16rem; --sidebar-width-icon: 3rem;"
  >
    <ul
      class="grid gap-4 grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 2xl:grid-cols-7 auto-rows-fr"
    >
      <li
        class="h-full"
      >
        <div
          class="relative h-full"
        >
          <div
            class="flex-none rounded-lg border bg-card text-card-foreground shadow-sm relative dark:shadow-slate-500/50 select-none flex flex-col overflow-hidden transition-all duration-300 ease-in-out hover:shadow-[0_8px_16px_-2px_var(--card-shadow-color)] h-full"
            style="--card-shadow-color: #00000040;"
          >
            <div
              style="content: ''; position: absolute; top: -3rem; right: -3rem; width: 16rem; height: 16rem; filter: blur(12px); z-index: 0;"
            />
            <div
              class="flex flex-col gap-2 flex-1 z-10 transition-all duration-300 ease-in-out p-4"
            >
              <div
                class="w-full flex justify-between items-center"
              >
                <div
                  class="flex items-center gap-2"
                >
                  <img
                    alt="Favicon"
                    class="rounded"
                    data-nimg="1"
                    decoding="async"
                    height="32"
                    loading="lazy"
                    src="/_next/image?url=https%3A%2F%2Fwww.google.com%2Fs2%2Ffavicons%3Fsz%3D32%26domain_url%3Dhttps%253A%252F%252Ftest.com&w=64&q=75"
                    srcset="/_next/image?url=https%3A%2F%2Fwww.google.com%2Fs2%2Ffavicons%3Fsz%3D32%26domain_url%3Dhttps%253A%252F%252Ftest.com&w=32&q=75 1x, /_next/image?url=https%3A%2F%2Fwww.google.com%2Fs2%2Ffavicons%3Fsz%3D32%26domain_url%3Dhttps%253A%252F%252Ftest.com&w=64&q=75 2x"
                    style="color: transparent; object-fit: cover;"
                    width="32"
                  />
                </div>
                <div
                  class="flex items-center gap-2"
                >
                  <div
                    class="flex items-center gap-2"
                  />
                  <button
                    aria-expanded="false"
                    aria-haspopup="menu"
                    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground h-8 w-8 p-0 z-30"
                    data-state="closed"
                    id="radix-:ra:"
                    type="button"
                  >
                    <span
                      class="sr-only"
                    >
                      Open menu
                    </span>
                    <svg
                      class="lucide lucide-ellipsis-vertical h-4 w-4"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <circle
                        cx="12"
                        cy="12"
                        r="1"
                      />
                      <circle
                        cx="12"
                        cy="5"
                        r="1"
                      />
                      <circle
                        cx="12"
                        cy="19"
                        r="1"
                      />
                    </svg>
                  </button>
                </div>
              </div>
              <div
                class="transition-all duration-300 ease-in-out origin-top flex flex-col flex-1"
              >
                <h3
                  class="tracking-tight text-sm font-medium transition-opacity duration-300"
                >
                  <a
                    class="before:absolute before:top-0 before:left-0 before:right-0 before:bottom-0 before:content-[''] before:z-10"
                    href="https://test.com"
                    rel="noopener noreferrer nofollow"
                    target="_blank"
                  >
                    Test App
                  </a>
                </h3>
                <div
                  class="flex flex-col flex-1 justify-between"
                >
                  <p
                    class="text-muted-foreground text-xs transition-all duration-300 ease-in-out max-h-[100px] opacity-100"
                  >
                    Test Description
                  </p>
                  <div
                    class="mt-auto pt-2 flex items-center gap-2 z-30"
                  >
                    <div
                      class="inline-flex items-center rounded-full px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-secondary-foreground hover:bg-secondary/80 bg-background/80 border border-muted"
                    >
                      test-tag
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </li>
      <li
        class="h-full"
      >
        <div
          class="relative h-full"
        >
          <div
            class="flex-none rounded-lg border bg-card text-card-foreground shadow-sm relative dark:shadow-slate-500/50 select-none flex flex-col overflow-hidden transition-all duration-300 ease-in-out hover:shadow-[0_8px_16px_-2px_var(--card-shadow-color)] h-full"
            style="--card-shadow-color: #00000040;"
          >
            <div
              style="content: ''; position: absolute; top: -3rem; right: -3rem; width: 16rem; height: 16rem; filter: blur(12px); z-index: 0;"
            />
            <div
              class="flex flex-col gap-2 flex-1 z-10 transition-all duration-300 ease-in-out p-4"
            >
              <div
                class="w-full flex justify-between items-center"
              >
                <div
                  class="flex items-center gap-2"
                >
                  <img
                    alt="Favicon"
                    class="rounded"
                    data-nimg="1"
                    decoding="async"
                    height="32"
                    loading="lazy"
                    src="/_next/image?url=https%3A%2F%2Fwww.google.com%2Fs2%2Ffavicons%3Fsz%3D32%26domain_url%3Dhttps%253A%252F%252Fprod.test.com&w=64&q=75"
                    srcset="/_next/image?url=https%3A%2F%2Fwww.google.com%2Fs2%2Ffavicons%3Fsz%3D32%26domain_url%3Dhttps%253A%252F%252Fprod.test.com&w=32&q=75 1x, /_next/image?url=https%3A%2F%2Fwww.google.com%2Fs2%2Ffavicons%3Fsz%3D32%26domain_url%3Dhttps%253A%252F%252Fprod.test.com&w=64&q=75 2x"
                    style="color: transparent; object-fit: cover;"
                    width="32"
                  />
                </div>
                <div
                  class="flex items-center gap-2"
                >
                  <div
                    class="flex items-center gap-2"
                  />
                  <button
                    aria-expanded="false"
                    aria-haspopup="menu"
                    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground h-8 w-8 p-0 z-30"
                    data-state="closed"
                    id="radix-:rf:"
                    type="button"
                  >
                    <span
                      class="sr-only"
                    >
                      Open menu
                    </span>
                    <svg
                      class="lucide lucide-ellipsis-vertical h-4 w-4"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <circle
                        cx="12"
                        cy="12"
                        r="1"
                      />
                      <circle
                        cx="12"
                        cy="5"
                        r="1"
                      />
                      <circle
                        cx="12"
                        cy="19"
                        r="1"
                      />
                    </svg>
                  </button>
                </div>
              </div>
              <div
                class="transition-all duration-300 ease-in-out origin-top flex flex-col flex-1"
              >
                <h3
                  class="tracking-tight text-sm font-medium transition-opacity duration-300"
                >
                  <a
                    class="before:absolute before:top-0 before:left-0 before:right-0 before:bottom-0 before:content-[''] before:z-10"
                    href="https://prod.test.com"
                    rel="noopener noreferrer nofollow"
                    target="_blank"
                  >
                    Multi Env App
                  </a>
                </h3>
                <div
                  class="flex flex-col flex-1 justify-between"
                >
                  <p
                    class="text-muted-foreground text-xs transition-all duration-300 ease-in-out max-h-[100px] opacity-100"
                  >
                    App with multiple environments
                  </p>
                  <div
                    class="mt-auto pt-2 flex items-center gap-2 z-30"
                  >
                    <div
                      class="inline-flex items-center rounded-full px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-secondary-foreground hover:bg-secondary/80 bg-background/80 border border-muted"
                    >
                      test-tag
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
